--- gcc.src/libiberty/pex-win32.c	2016-06-29 09:49:11.149989000 +0200
+++ gcc-5/libiberty/pex-win32.c	2016-06-29 15:20:54.584575900 +0200
@@ -356,7 +356,7 @@
       needs_quotes = 0;
       for (j = 0; argv[i][j]; j++)
 	{
-	  if (argv[i][j] == ' ' || argv[i][j] == '\t' || argv[i][j] == '"')
+	  if (argv[i][j] == ' ' || argv[i][j] == '\t' || argv[i][j] == '"' || argv[i][j] == '@')
 	    {
 	      needs_quotes = 1;
 	    }
@@ -388,7 +388,7 @@
       needs_quotes = 0;
       for (j = 0; argv[i][j]; j++)
         {
-          if (argv[i][j] == ' ' || argv[i][j] == '\t' || argv[i][j] == '"')
+          if (argv[i][j] == ' ' || argv[i][j] == '\t' || argv[i][j] == '"' || argv[i][j] == '@')
             {
               needs_quotes = 1;
               break;
