diff -u -r clang-4.0.orig/CMakeLists.txt clang-4.0/CMakeLists.txt
--- clang-4.0.orig/CMakeLists.txt	2017-03-23 20:17:39.000000000 +0100
+++ clang-4.0/CMakeLists.txt	2017-07-26 14:30:17.617930435 +0200
@@ -301,7 +301,7 @@
 
 # They are used as destination of target generators.
 set(LLVM_RUNTIME_OUTPUT_INTDIR ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/bin)
-set(LLVM_LIBRARY_OUTPUT_INTDIR ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/lib${LLVM_LIBDIR_SUFFIX})
+set(LLVM_LIBRARY_OUTPUT_INTDIR ${CMAKE_CURRENT_BINARY_DIR}/Release/lib${LLVM_LIBDIR_SUFFIX})
 if(WIN32 OR CYGWIN)
   # DLL platform -- put DLLs into bin.
   set(LLVM_SHLIB_OUTPUT_INTDIR ${LLVM_RUNTIME_OUTPUT_INTDIR})
diff -u -r clang-4.0.orig/lib/Support/Windows/Host.inc clang-4.0/lib/Support/Windows/Host.inc
--- clang-4.0.orig/lib/Support/Windows/Host.inc	2014-09-24 06:45:14.000000000 +0200
+++ clang-4.0/lib/Support/Windows/Host.inc	2017-07-26 14:30:17.618930418 +0200
@@ -18,5 +18,5 @@
 using namespace llvm;
 
 std::string sys::getDefaultTargetTriple() {
-  return Triple::normalize(LLVM_DEFAULT_TARGET_TRIPLE);
+  return LLVM_DEFAULT_TARGET_TRIPLE;
 }
diff -u -r clang-4.0.orig/tools/clang/include/clang/Driver/Driver.h clang-4.0/tools/clang/include/clang/Driver/Driver.h
--- clang-4.0.orig/tools/clang/include/clang/Driver/Driver.h	2016-11-18 12:18:28.000000000 +0100
+++ clang-4.0/tools/clang/include/clang/Driver/Driver.h	2017-07-26 14:30:17.618930418 +0200
@@ -147,6 +147,9 @@
   /// Dynamic loader prefix, if present
   std::string DyldPrefix;
 
+  /// Default GCC install path
+  std::string DefaultGCCInstallPath;
+
   /// If the standard library is used
   bool UseStdLib;
 
@@ -285,6 +288,11 @@
   const std::string &getTitle() { return DriverTitle; }
   void setTitle(std::string Value) { DriverTitle = std::move(Value); }
 
+  /// \brief Get the default path of GCC install prefix
+  const char *getDefaultGCCInstallPath() const {
+    return DefaultGCCInstallPath.c_str();
+  }
+
   /// \brief Get the path to the main clang executable.
   const char *getClangProgramPath() const {
     return ClangExecutable.c_str();
diff -u -r clang-4.0.orig/tools/clang/lib/Driver/Driver.cpp clang-4.0/tools/clang/lib/Driver/Driver.cpp
--- clang-4.0.orig/tools/clang/lib/Driver/Driver.cpp	2017-01-05 06:20:27.000000000 +0100
+++ clang-4.0/tools/clang/lib/Driver/Driver.cpp	2017-07-26 14:31:48.656383110 +0200
@@ -1,3 +1,4 @@
+
 //===--- Driver.cpp - Clang GCC Compatible Driver -------------------------===//
 //
 //                     The LLVM Compiler Infrastructure
@@ -27,6 +28,7 @@
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/ADT/StringSet.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "llvm/Option/Arg.h"
 #include "llvm/Option/ArgList.h"
 #include "llvm/Option/OptSpecifier.h"
@@ -50,13 +52,22 @@
 using namespace clang;
 using namespace llvm::opt;
 
+static size_t getPrefixFolderOf(StringRef prefix, StringRef path) {
+  size_t Common;
+  for (Common = 0; Common < prefix.size() && Common < path.size()
+                && (prefix[Common] == path[Common] || (llvm::sys::path::is_separator(prefix[Common])
+                                                    && llvm::sys::path::is_separator(path[Common]))); ++Common) { }
+  return Common;
+}
+
 Driver::Driver(StringRef ClangExecutable, StringRef DefaultTargetTriple,
                DiagnosticsEngine &Diags,
                IntrusiveRefCntPtr<vfs::FileSystem> VFS)
     : Opts(createDriverOptTable()), Diags(Diags), VFS(std::move(VFS)),
       Mode(GCCMode), SaveTemps(SaveTempsNone), BitcodeEmbed(EmbedNone),
       LTOMode(LTOK_None), ClangExecutable(ClangExecutable),
-      SysRoot(DEFAULT_SYSROOT), UseStdLib(true),
+      SysRoot(DEFAULT_SYSROOT), DefaultGCCInstallPath(GCC_INSTALL_PREFIX),
+      UseStdLib(true),
       DriverTitle("clang LLVM compiler"), CCPrintOptionsFilename(nullptr),
       CCPrintHeadersFilename(nullptr), CCLogDiagnosticsFilename(nullptr),
       CCCPrintBindings(false), CCPrintHeaders(false), CCLogDiagnostics(false),
@@ -72,6 +83,47 @@
   Dir = llvm::sys::path::parent_path(ClangExecutable);
   InstalledDir = Dir; // Provide a sensible default installed dir.
 
+  StringRef DefaultSysroot = DEFAULT_SYSROOT;
+  StringRef DefaultGCCInstallPathRef = GCC_INSTALL_PREFIX;
+  if (!DefaultSysroot.empty() || !DefaultGCCInstallPathRef.empty()) {
+    // Try to relocate the sysroot/gcc install path when prefix of
+    // ClangExecutable not equal to LLVM_PREFIX.
+    StringRef LLVMPrefix = LLVM_PREFIX;
+    size_t CommonPrefixLength = getPrefixFolderOf(LLVMPrefix, ClangExecutable);
+    if (CommonPrefixLength != LLVMPrefix.size()) {
+      StringRef Sep = "/";
+      {
+        size_t CommonSysrootLength = getPrefixFolderOf(LLVMPrefix, DefaultSysroot);
+        StringRef RelativePath = DefaultSysroot.substr(CommonSysrootLength);
+        StringRef SkipPath = LLVMPrefix.substr(CommonSysrootLength + (llvm::sys::path::is_separator(LLVMPrefix[CommonSysrootLength]) ? 1 : 0));
+        StringRef RelocatedBase = llvm::sys::path::parent_path(Dir);
+        for (llvm::sys::path::const_iterator it = llvm::sys::path::begin(SkipPath), end = llvm::sys::path::end(SkipPath);
+             it != end;
+             ++it) {
+          RelocatedBase = llvm::sys::path::parent_path(RelocatedBase);
+        }
+        if (llvm::sys::fs::is_directory(RelocatedBase + Sep + RelativePath)) {
+          SysRoot =  (RelocatedBase + Sep + RelativePath).str();
+        }
+      }
+
+      {
+        size_t CommonGCCInstallLength = getPrefixFolderOf(LLVMPrefix, DefaultGCCInstallPath);
+        StringRef RelativePath = DefaultGCCInstallPath.substr(CommonGCCInstallLength);
+        StringRef SkipPath = LLVMPrefix.substr(CommonGCCInstallLength + (llvm::sys::path::is_separator(LLVMPrefix[CommonGCCInstallLength]) ? 1 : 0));
+        StringRef RelocatedBase = llvm::sys::path::parent_path(Dir);
+        for (llvm::sys::path::const_iterator it = llvm::sys::path::begin(SkipPath);
+             it != llvm::sys::path::end(SkipPath);
+             ++it) {
+          RelocatedBase = llvm::sys::path::parent_path(RelocatedBase);
+        }
+        if (llvm::sys::fs::is_directory(RelocatedBase + Sep + RelativePath)) {
+          DefaultGCCInstallPath =  (RelocatedBase + Sep + RelativePath).str();
+        }
+      }
+    }
+  }
+
   // Compute the path to the resource directory.
   StringRef ClangResourceDir(CLANG_RESOURCE_DIR);
   SmallString<128> P(Dir);
diff -u -r clang-4.0.orig/tools/clang/lib/Driver/ToolChain.cpp clang-4.0/tools/clang/lib/Driver/ToolChain.cpp
--- clang-4.0.orig/tools/clang/lib/Driver/ToolChain.cpp	2016-12-14 17:46:50.000000000 +0100
+++ clang-4.0/tools/clang/lib/Driver/ToolChain.cpp	2017-07-26 14:32:45.174416335 +0200
@@ -16,6 +16,7 @@
 #include "clang/Driver/DriverDiagnostic.h"
 #include "clang/Driver/Options.h"
 #include "clang/Driver/SanitizerArgs.h"
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "llvm/ADT/SmallString.h"
 #include "llvm/Option/Arg.h"
 #include "llvm/Option/ArgList.h"
diff -u -r clang-4.0.orig/tools/clang/lib/Driver/ToolChains.cpp clang-4.0/tools/clang/lib/Driver/ToolChains.cpp
--- clang-4.0.orig/tools/clang/lib/Driver/ToolChains.cpp	2017-05-31 12:01:14.000000000 +0200
+++ clang-4.0/tools/clang/lib/Driver/ToolChains.cpp	2017-07-26 14:30:17.624930316 +0200
@@ -13,6 +13,7 @@
 #include "clang/Basic/Version.h"
 #include "clang/Basic/VirtualFileSystem.h"
 #include "clang/Config/config.h" // for GCC_INSTALL_PREFIX
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "clang/Driver/Compilation.h"
 #include "clang/Driver/Distro.h"
 #include "clang/Driver/Driver.h"
@@ -1388,11 +1389,11 @@
   return false;
 }
 
-static llvm::StringRef getGCCToolchainDir(const ArgList &Args) {
+static StringRef getGCCToolchainDir(const ArgList &Args, const Driver &D) {
   const Arg *A = Args.getLastArg(options::OPT_gcc_toolchain);
   if (A)
     return A->getValue();
-  return GCC_INSTALL_PREFIX;
+  return D.getDefaultGCCInstallPath();
 }
 
 /// \brief Initialize a GCCInstallationDetector from the driver.
@@ -1423,7 +1424,7 @@
   SmallVector<std::string, 8> Prefixes(D.PrefixDirs.begin(),
                                        D.PrefixDirs.end());
 
-  StringRef GCCToolchainDir = getGCCToolchainDir(Args);
+  StringRef GCCToolchainDir = getGCCToolchainDir(Args, D);
   if (GCCToolchainDir != "") {
     if (GCCToolchainDir.back() == '/')
       GCCToolchainDir = GCCToolchainDir.drop_back(); // remove the /
