diff -Naur clang-3.4.src/tools/clang/include/clang/Driver/Driver.h clang-3.4/tools/clang/include/clang/Driver/Driver.h
--- clang-3.4.src/tools/clang/include/clang/Driver/Driver.h	2014-04-11 22:31:22.000000000 +0200
+++ clang-3.4/tools/clang/include/clang/Driver/Driver.h	2016-05-31 10:46:51.559145400 +0200
@@ -97,6 +97,9 @@
   /// Dynamic loader prefix, if present
   std::string DyldPrefix;
 
+  /// Default GCC install path
+  std::string DefaultGCCInstallPath;
+
   /// If the standard library is used
   bool UseStdLib;
 
@@ -214,6 +217,11 @@
   const std::string &getTitle() { return DriverTitle; }
   void setTitle(std::string Value) { DriverTitle = Value; }
 
+  /// \brief Get the default path of GCC install prefix
+  const char *getDefaultGCCInstallPath() const {
+    return DefaultGCCInstallPath.c_str();
+  }
+
   /// \brief Get the path to the main clang executable.
   const char *getClangProgramPath() const {
     return ClangExecutable.c_str();
diff -Naur clang-3.4.src/tools/clang/lib/Driver/Driver.cpp clang-3.4/tools/clang/lib/Driver/Driver.cpp
--- clang-3.4.src/tools/clang/lib/Driver/Driver.cpp	2014-04-21 22:30:27.000000000 +0200
+++ clang-3.4/tools/clang/lib/Driver/Driver.cpp	2016-05-31 15:08:34.470731800 +0200
@@ -39,18 +39,28 @@
 
 // FIXME: It would prevent us from including llvm-config.h
 // if config.h were included before system_error.h.
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "clang/Config/config.h"
 
 using namespace clang::driver;
 using namespace clang;
 using namespace llvm::opt;
 
+static size_t getPrefixFolderOf(StringRef prefix, StringRef path) {
+  size_t Common;
+  for (Common = 0; Common < prefix.size() && Common < path.size()
+                && (prefix[Common] == path[Common] || (llvm::sys::path::is_separator(prefix[Common])
+                                                    && llvm::sys::path::is_separator(path[Common]))); ++Common) { }
+  return Common;
+}
+
 Driver::Driver(StringRef ClangExecutable,
                StringRef DefaultTargetTriple,
                StringRef DefaultImageName,
                DiagnosticsEngine &Diags)
   : Opts(createDriverOptTable()), Diags(Diags), Mode(GCCMode),
     ClangExecutable(ClangExecutable), SysRoot(DEFAULT_SYSROOT),
+    DefaultGCCInstallPath(GCC_INSTALL_PREFIX),
     UseStdLib(true), DefaultTargetTriple(DefaultTargetTriple),
     DefaultImageName(DefaultImageName),
     DriverTitle("clang LLVM compiler"),
@@ -63,6 +73,47 @@
 
   Name = llvm::sys::path::stem(ClangExecutable);
   Dir  = llvm::sys::path::parent_path(ClangExecutable);
+ 
+  StringRef DefaultSysroot = DEFAULT_SYSROOT;
+  StringRef DefaultGCCInstallPathRef = GCC_INSTALL_PREFIX;
+  if (!DefaultSysroot.empty() || !DefaultGCCInstallPathRef.empty()) {
+    // Try to relocate the sysroot/gcc install path when prefix of
+    // ClangExecutable not equal to LLVM_PREFIX.
+    StringRef LLVMPrefix = LLVM_PREFIX;
+    size_t CommonPrefixLength = getPrefixFolderOf(LLVMPrefix, ClangExecutable);
+    if (CommonPrefixLength != LLVMPrefix.size()) {
+      StringRef Sep = "/";
+      {
+        size_t CommonSysrootLength = getPrefixFolderOf(LLVMPrefix, DefaultSysroot);
+        StringRef RelativePath = DefaultSysroot.substr(CommonSysrootLength);
+        StringRef SkipPath = LLVMPrefix.substr(CommonSysrootLength + (llvm::sys::path::is_separator(LLVMPrefix[CommonSysrootLength]) ? 1 : 0));
+        StringRef RelocatedBase = llvm::sys::path::parent_path(Dir);
+        for (llvm::sys::path::const_iterator it = llvm::sys::path::begin(SkipPath), end = llvm::sys::path::end(SkipPath);
+             it != end;
+             ++it) {
+          RelocatedBase = llvm::sys::path::parent_path(RelocatedBase);
+        }
+        if (llvm::sys::fs::is_directory(RelocatedBase + Sep + RelativePath)) {
+          SysRoot =  (RelocatedBase + Sep + RelativePath).str();
+        }
+      }
+
+      {
+        size_t CommonGCCInstallLength = getPrefixFolderOf(LLVMPrefix, DefaultGCCInstallPath);
+        StringRef RelativePath = DefaultGCCInstallPath.substr(CommonGCCInstallLength);
+        StringRef SkipPath = LLVMPrefix.substr(CommonGCCInstallLength + (llvm::sys::path::is_separator(LLVMPrefix[CommonGCCInstallLength]) ? 1 : 0));
+        StringRef RelocatedBase = llvm::sys::path::parent_path(Dir);
+        for (llvm::sys::path::const_iterator it = llvm::sys::path::begin(SkipPath);
+             it != llvm::sys::path::end(SkipPath);
+             ++it) {
+          RelocatedBase = llvm::sys::path::parent_path(RelocatedBase);
+        }
+        if (llvm::sys::fs::is_directory(RelocatedBase + Sep + RelativePath)) {
+          DefaultGCCInstallPath =  (RelocatedBase + Sep + RelativePath).str();
+        }
+      }
+    }
+  }
 
   // Compute the path to the resource directory.
   StringRef ClangResourceDir(CLANG_RESOURCE_DIR);
diff -Naur clang-3.4.src/tools/clang/lib/Driver/ToolChains.cpp clang-3.4/tools/clang/lib/Driver/ToolChains.cpp
--- clang-3.4.src/tools/clang/lib/Driver/ToolChains.cpp	2014-04-26 01:38:13.000000000 +0200
+++ clang-3.4/tools/clang/lib/Driver/ToolChains.cpp	2016-05-31 10:46:51.575146100 +0200
@@ -33,6 +33,7 @@
 
 // FIXME: This needs to be listed last until we fix the broken include guards
 // in these files and the LLVM config.h files.
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "clang/Config/config.h" // for GCC_INSTALL_PREFIX
 
 #include <cstdlib> // ::getenv
@@ -1007,11 +1008,11 @@
   return false;
 }
 
-static StringRef getGCCToolchainDir(const ArgList &Args) {
+static StringRef getGCCToolchainDir(const ArgList &Args, const Driver &D) {
   const Arg *A = Args.getLastArg(options::OPT_gcc_toolchain);
   if (A)
     return A->getValue();
-  return GCC_INSTALL_PREFIX;
+  return D.getDefaultGCCInstallPath();
 }
 
 /// \brief Construct a GCCInstallationDetector from the driver.
@@ -1043,7 +1044,7 @@
   SmallVector<std::string, 8> Prefixes(D.PrefixDirs.begin(),
                                        D.PrefixDirs.end());
 
-  StringRef GCCToolchainDir = getGCCToolchainDir(Args);
+  StringRef GCCToolchainDir = getGCCToolchainDir(Args, D);
   if (GCCToolchainDir != "") {
     if (GCCToolchainDir.back() == '/')
       GCCToolchainDir = GCCToolchainDir.drop_back(); // remove the /
