diff -ur llvm-3.5.0.orig/lib/Analysis/ConstantFolding.cpp llvm-3.5.0.src/lib/Analysis/ConstantFolding.cpp
--- llvm-3.5.0.orig/lib/Analysis/ConstantFolding.cpp	2014-07-14 22:39:26 +0000
+++ llvm-3.5.0.src/lib/Analysis/ConstantFolding.cpp	2014-11-21 09:57:47 +0000
@@ -38,6 +38,7 @@
 #include <cmath>
 
 #ifdef HAVE_FENV_H
+#define _GLIBCXX_HAVE_FENV_H 1
 #include <fenv.h>
 #endif
 
diff -ur llvm-3.5.0.orig/tools/clang/include/clang/Driver/Driver.h llvm-3.5.0.src/tools/clang/include/clang/Driver/Driver.h
--- llvm-3.5.0.orig/tools/clang/include/clang/Driver/Driver.h	2014-05-15 22:26:36 +0000
+++ llvm-3.5.0.src/tools/clang/include/clang/Driver/Driver.h	2014-11-21 08:34:21 +0000
@@ -97,6 +97,9 @@
   /// Dynamic loader prefix, if present
   std::string DyldPrefix;
 
+  /// Default GCC install path
+  std::string DefaultGCCInstallPath;
+
   /// If the standard library is used
   bool UseStdLib;
 
@@ -213,6 +216,11 @@
   const std::string &getTitle() { return DriverTitle; }
   void setTitle(std::string Value) { DriverTitle = Value; }
 
+  /// \brief Get the default path of GCC install prefix
+  const char *getDefaultGCCInstallPath() const {
+    return DefaultGCCInstallPath.c_str();
+  }
+
   /// \brief Get the path to the main clang executable.
   const char *getClangProgramPath() const {
     return ClangExecutable.c_str();
diff -ur llvm-3.5.0.orig/tools/clang/lib/Driver/Driver.cpp llvm-3.5.0.src/tools/clang/lib/Driver/Driver.cpp
--- llvm-3.5.0.orig/tools/clang/lib/Driver/Driver.cpp	2014-07-11 14:28:45 +0000
+++ llvm-3.5.0.src/tools/clang/lib/Driver/Driver.cpp	2014-11-21 08:39:41 +0000
@@ -24,6 +24,7 @@
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/ADT/StringSet.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "llvm/Option/Arg.h"
 #include "llvm/Option/ArgList.h"
 #include "llvm/Option/OptSpecifier.h"
@@ -44,11 +45,20 @@
 using namespace clang;
 using namespace llvm::opt;
 
+static bool isPrefixFolderOf(StringRef prefix, StringRef path) {
+  bool IsEndWithBackslash = ((path.size() >= prefix.size()) &&
+                              path[prefix.size()] == '/') ||
+                            prefix[prefix.size()-1] == '/';
+  bool IsPrefix = path.startswith(prefix);
+  return (IsEndWithBackslash && IsPrefix) || prefix.equals(path);
+}
+
 Driver::Driver(StringRef ClangExecutable,
                StringRef DefaultTargetTriple,
                DiagnosticsEngine &Diags)
   : Opts(createDriverOptTable()), Diags(Diags), Mode(GCCMode),
     ClangExecutable(ClangExecutable), SysRoot(DEFAULT_SYSROOT),
+    DefaultGCCInstallPath(GCC_INSTALL_PREFIX),
     UseStdLib(true), DefaultTargetTriple(DefaultTargetTriple),
     DefaultImageName("a.out"),
     DriverTitle("clang LLVM compiler"),
@@ -59,9 +69,38 @@
     CCGenDiagnostics(false), CCCGenericGCCName(""), CheckInputsExist(true),
     CCCUsePCH(true), SuppressMissingInputWarning(false) {
 
+  {
+  }
   Name = llvm::sys::path::stem(ClangExecutable);
   Dir  = llvm::sys::path::parent_path(ClangExecutable);
 
+  StringRef DefaultSysroot = DEFAULT_SYSROOT;
+  StringRef DefaultGCCInstallPathRef = GCC_INSTALL_PREFIX;
+  if (!DefaultSysroot.empty() || !DefaultGCCInstallPathRef.empty()) {
+    // Try to relocate the sysroot/gcc install path when prefix of
+    // ClangExecutable not equal to LLVM_PREFIX.
+    StringRef LLVMPrefix = LLVM_PREFIX;
+    if (!isPrefixFolderOf(LLVMPrefix, ClangExecutable)) {
+      // We only recognize the case for default sysroot/gcc install path is
+      // sub-folder of install prefix.
+      StringRef UpperFolder = "/../";
+      Twine RelocatedBase = Dir + UpperFolder;
+      if (isPrefixFolderOf(LLVMPrefix, SysRoot)) {
+        // Need relocate sysroot !
+        StringRef RelativePath = DefaultSysroot.data() + LLVMPrefix.size();
+        Twine RelocatedPath = RelocatedBase + RelativePath;
+        SysRoot = RelocatedPath.str();
+      }
+      if (isPrefixFolderOf(LLVMPrefix, DefaultGCCInstallPath)) {
+        // Need relocate default gcc install path !
+        StringRef RelativePath = DefaultGCCInstallPathRef.data() +
+                                 LLVMPrefix.size();
+        Twine RelocatedPath = RelocatedBase + RelativePath;
+        DefaultGCCInstallPath = RelocatedPath.str();
+      }
+    }
+  }
+
   // Compute the path to the resource directory.
   StringRef ClangResourceDir(CLANG_RESOURCE_DIR);
   SmallString<128> P(Dir);
diff -ur llvm-3.5.0.orig/tools/clang/lib/Driver/ToolChain.cpp llvm-3.5.0.src/tools/clang/lib/Driver/ToolChain.cpp
--- llvm-3.5.0.orig/tools/clang/lib/Driver/ToolChain.cpp	2014-06-26 14:23:45 +0000
+++ llvm-3.5.0.src/tools/clang/lib/Driver/ToolChain.cpp	2014-11-21 08:40:52 +0000
@@ -17,6 +17,7 @@
 #include "clang/Driver/ToolChain.h"
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "llvm/Option/Arg.h"
 #include "llvm/Option/ArgList.h"
 #include "llvm/Option/Option.h"
diff -ur llvm-3.5.0.orig/tools/clang/lib/Driver/ToolChains.cpp llvm-3.5.0.src/tools/clang/lib/Driver/ToolChains.cpp
--- llvm-3.5.0.orig/tools/clang/lib/Driver/ToolChains.cpp	2014-08-27 20:07:31 +0000
+++ llvm-3.5.0.src/tools/clang/lib/Driver/ToolChains.cpp	2014-11-21 08:44:06 +0000
@@ -11,6 +11,7 @@
 #include "clang/Basic/ObjCRuntime.h"
 #include "clang/Basic/Version.h"
 #include "clang/Config/config.h" // for GCC_INSTALL_PREFIX
+#include "llvm/Config/config.h" // for LLVM_PREFIX
 #include "clang/Driver/Compilation.h"
 #include "clang/Driver/Driver.h"
 #include "clang/Driver/DriverDiagnostic.h"
@@ -1197,11 +1198,11 @@
   return false;
 }
 
-static llvm::StringRef getGCCToolchainDir(const ArgList &Args) {
+static StringRef getGCCToolchainDir(const ArgList &Args, const Driver &D) {
   const Arg *A = Args.getLastArg(options::OPT_gcc_toolchain);
   if (A)
     return A->getValue();
-  return GCC_INSTALL_PREFIX;
+  return D.getDefaultGCCInstallPath();
 }
 
 /// \brief Initialize a GCCInstallationDetector from the driver.
@@ -1232,7 +1233,7 @@
   SmallVector<std::string, 8> Prefixes(D.PrefixDirs.begin(),
                                        D.PrefixDirs.end());
 
-  StringRef GCCToolchainDir = getGCCToolchainDir(Args);
+  StringRef GCCToolchainDir = getGCCToolchainDir(Args, D);
   if (GCCToolchainDir != "") {
     if (GCCToolchainDir.back() == '/')
       GCCToolchainDir = GCCToolchainDir.drop_back(); // remove the /
diff -ur llvm-3.5.0.orig/lib/Support/Windows/Host.inc llvm-3.5.0.src/lib/Support/Windows/Host.inc
--- llvm-3.5.0.orig/lib/Support/Windows/Host.inc	2014-03-29 02:08:53.000000000 +0100
+++ llvm-3.5.0.src/lib/Support/Windows/Host.inc	2016-05-24 11:34:53.806478400 +0200
@@ -18,5 +18,5 @@
 using namespace llvm;
 
 std::string sys::getDefaultTargetTriple() {
-  return Triple::normalize(LLVM_DEFAULT_TARGET_TRIPLE);
+  return LLVM_DEFAULT_TARGET_TRIPLE;
 }
